app/controllers/upload.controller.js

```diff
@@ -0,0 +1,262 @@
+		res.status(400).send({
+			message: "Invalid signature."
+		});
+		return;
+	}
+
+	let files = req.body.files;
+	if(typeof files === "undefined") {
+		res.status(400).send({
+			message: "Missing files field."
+		});
+		return;
+	}
+	if(typeof files !== "object" || !Array.isArray(files)) {
+		res.status(400).send({
+			message: "Invalid files field."
+		});
+		return;
+	}
+	if(files.length == 0) {
+		res.status(400).send({
+			message: "Empty files field."
+		});
+		return;
+	}
+
+	if(files.length > 64) {
+		res.status(400).send({
+			message: "Too many files. Max 64."
+		});
+		return;
+	}
+
+	for(let i = 0; i < files.length; i++) {
+		if(typeof files[i] !== "string") {
+			res.status(400).send({
+				message: "Invalid files field."
+			});
+			return;
+		}
+		// TODO: validate URL format better
+		if(!files[i].startsWith('http://') && !files[i].startsWith('https://') && !files[i].startsWith('ipfs://')) {
+			res.status(400).send({
+				message: "Invalid files URI."
+			});
+			return;
+		}
+	}
+
+	// validate quote
+	Quote.get(quoteId, async (err, quote) => {
+		if(err) {
+			if(err.code == 404) {
+				res.status(404).send({
+					message: "Quote not found"
+				});
+				return;
+			}
+			res.status(500).send({
+				message:
+					err.message || "Error occurred while validating quote."
+			});
+		}
+
+		// see if token still accepted
+		const paymentToken = acceptToken(quote.chainId, quote.tokenAddress);
+		if(!paymentToken) {
+			res.status(400).send({
+				message: "Payment token no longer accepted."
+			});
+			return;
+		}
+
+		// check status of quote
+		if(quote.status != Quote.QUOTE_STATUS_WAITING) {
+			if(quote.status == Quote.QUOTE_STATUS_UPLOAD_END) {
+				res.status(400).send({
+					message: "Quote has been completed."
+				});
+				return;
+			}
+			else {
+				res.status(400).send({
+					message: "Quote is being processed."
+				});
+				return;
+			}
+		}
+
+		// check if new price is sufficient
+		let bundlr;
+		try {
+			bundlr = new Bundlr.default(process.env.ARWEAVE_GATEWAY_URI, paymentToken.name, process.env.PRIVATE_KEY, paymentToken.providerUrl ? {providerUrl: paymentToken.providerUrl, contractAddress: paymentToken.tokenAddress} : {});
+		}
+		catch(err) {
+			res.status(500).send({
+				message: err.message
+			});
+			return;
+		}	
+
+		//console.log(`Whole quote size: ${quote.size}`);
+
+		const priceWei = await bundlr.getPrice(quote.size);
+		const tokenAmount = bundlr.utils.unitConverter(priceWei);
+
+		if(parseFloat(tokenAmount) > parseFloat(quote.tokenAmount)) {
+			res.status(402).send({
+				message: `Quoted tokenAmount is less than current rate. Quoted amount: ${quote.tokenAmount}, current rate: ${tokenAmount}`
+			});
+			return;
+		}
+
+		res.send(null); // send 200
+
+		// change status
+		Quote.setStatus(quoteId, Quote.QUOTE_STATUS_PAYMENT_START);
+
+		// TODO: Pull WETH from user's account into our EOA using transferFrom(userAddress, amount)
+		// TODO: Unwrap WETH to ETH
```

Comment:
If the server dies somewhere while processing the payment, we'll need a way to restart it on init.

Reply:
Created issue to address this: https://github.com/MantisClone/arweave-upload/issues/21